### PR TITLE
Pass flags to `rustdoc` shim without env. vars

### PR DESCRIPTION
**Pull request by Kobzol**
Discussed here: https://github.com/rust-lang/rust/pull/116448#issuecomment-1748785961. Since it was not really documented why these flags were passed through the shim, I guess that the only way to find out if it's really needed... is to remove it :)

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
I recalled what the issue with doctests was - bootstrap passes options to rustdoc only when it calls rustdoc directly, not through cargo.
The options should also be passed when rustdoc is run through cargo (doctests are tested that way).

**Comment by petrochenkov** [comment]
This really needs to be tested separately, with full `x.py test --stage 2` https://github.com/rust-lang/rust/pull/116448#issuecomment-1748940140.
It's also quite possible that it is broken on master.

**Comment by Kobzol** [comment]
> I recalled what the issue with doctests was - bootstrap passes options to rustdoc only when it calls rustdoc directly, not through cargo. The options should also be passed when rustdoc is run through cargo (doctests are tested that way).

Yes, this accurately describes the state on `master`.

Do you have an example of a `x.py test` command or a subdirectory of `tests` where doc tests are executed through cargo? I tried using `x.py test --doc`, but couldn't really find a place where it would be applied. It seems to me that this only happens when we test tools and submodules (?).

**Comment by petrochenkov** [comment]
`cargo test` on standard library crates or rustc crates should run rustdoc with doctest building & running?
I don't know how to run it in a more fine-grained way, I just remember that it was included into full `x.py test` (at least back in 2017 :D).

**Comment by Kobzol** [comment]
I looked around and there are indeed a few places where rustdoc is being executed without the linker being set (`CargoTest`, `Bootstrap` at least). But this is a bigger issue I think - there are some places in bootstrap where we execute cargo that doesn't go through the main Cargo builder function. This means that there might be more places where the linker isn't set.

If we set `use-lld` or `linker`, should it mean that *absolutely all* relevant commands executed by bootstrap should change the linker/flavor? If yes, then this might need some more restructuring, rather than just adding it arbitrarily to a few places where rustdoc is invoked by cargo.

**Comment by petrochenkov** [comment]
>If we set `use-lld` or `linker`, should it mean that *absolutely all* relevant commands executed by bootstrap should change the linker/flavor?

Yes, that was the case when I originally implemented `linker` and full `use-lld` support.
If they are not used in some cases it's a later regression.

**Comment by Kobzol** [comment]
Okay so I investigated a bit and the only place where rustdoc is invoked manually (outside of `Builder::rustdoc_cmd`) is here: https://github.com/rust-lang/rust/blob/acd1869c8f0cfe3734c5dc3afc72133cd7d56bb9/src/bootstrap/test.rs#L846. There's an issue that here we need to prepend the `rustdoc-theme` command *before* actually passing the rustdoc command, but I don't think that `Command` allows prepending arguments, so that we could call `rustdoc_cmd` and then prepend `rustdoc-theme`.

**Comment by Kobzol** [comment]
I have added two functions to add the LLD flags to both normal rustdoc and cargo rustdoc invocations. It's far from pretty..

**Comment by Kobzol** [comment]
Rebased and tested full `x.py test --stage 2` locally.

@rustbot ready

File: src/bootstrap/src/utils/helpers.rs [hunk 1]
```diff
@@ -471,21 +472,63 @@ pub fn extract_beta_rev(version: &str) -> Option<String> {
     count
 }
 
+pub enum LldThreadMode {
+    SingleThreaded,
+    MultiThreaded,
+}
```
Comment (petrochenkov) [inline src/bootstrap/src/utils/helpers.rs:478]: I'd personally keep this a `bool`, but even with enum I'd select something less verbose, like `LldThreads::{Yes,No}` or something.

File: src/bootstrap/src/utils/helpers.rs [hunk 2]
```diff
@@ -471,21 +472,63 @@ pub fn extract_beta_rev(version: &str) -> Option<String> {
...
     target: TargetSelection,
-    single_threaded: bool,
+    lld_thread_mode: LldThreadMode,
 ) {
+    let args = build_rustdoc_lld_flags(builder, target, lld_thread_mode);
+    let mut flags = cmd
+        .get_envs()
+        .find(|(k, _)| *k == OsStr::new("RUSTDOCFLAGS"))
+        .and_then(|(_, v)| v)
+        .unwrap_or(OsStr::new(""))
```
Comment (petrochenkov) [inline src/bootstrap/src/utils/helpers.rs:501]:
```suggestion
        .unwrap_or_default()
```

File: src/bootstrap/src/utils/helpers.rs [hunk 3]
```diff
@@ -471,21 +472,63 @@ pub fn extract_beta_rev(version: &str) -> Option<String> {
...
+
 pub fn add_rustdoc_lld_flags(
     cmd: &mut Command,
-    builder: &Builder,
+    builder: &Builder<'_>,
+    target: TargetSelection,
+    lld_thread_mode: LldThreadMode,
+) {
+    let args = build_rustdoc_lld_flags(builder, target, lld_thread_mode);
+    cmd.args(args);
```
Comment (petrochenkov) [inline src/bootstrap/src/utils/helpers.rs:487]:
```suggestion
    cmd.args(build_rustdoc_lld_flags(builder, target, lld_thread_mode));
```

File: src/bootstrap/src/utils/helpers.rs [hunk 4]
```diff
@@ -471,21 +472,63 @@ pub fn extract_beta_rev(version: &str) -> Option<String> {
...
+    builder: &Builder<'_>,
     target: TargetSelection,
-    single_threaded: bool,
+    lld_thread_mode: LldThreadMode,
 ) {
+    let args = build_rustdoc_lld_flags(builder, target, lld_thread_mode);
+    let mut flags = cmd
+        .get_envs()
+        .find(|(k, _)| *k == OsStr::new("RUSTDOCFLAGS"))
+        .and_then(|(_, v)| v)
```
Comment (petrochenkov) [inline src/bootstrap/src/utils/helpers.rs:500]: Maybe `find_map` (with `flatten`?) will look better here?

File: src/bootstrap/src/core/builder.rs [hunk 5]
```diff
@@ -18,7 +18,10 @@ use crate::core::build_steps::{check, clean, compile, dist, doc, install, run, s
 use crate::core::config::flags::{Color, Subcommand};
 use crate::core::config::{DryRun, SplitDebuginfo, TargetSelection};
 use crate::utils::cache::{Cache, Interned, INTERNER};
-use crate::utils::helpers::{self, add_dylib_path, add_link_lib_path, exe, libdir, output, t};
+use crate::utils::helpers::{
+    self, add_dylib_path, add_link_lib_path, add_rustdoc_lld_flags, exe, libdir, output, t,
+    LldThreadMode,
+};
```
Comment (petrochenkov) [inline src/bootstrap/src/core/builder.rs:24]: Nit: could you keep the single-line import style?

**Comment by petrochenkov** [comment]
I'm also not sure whether the CI failure in https://github.com/rust-lang/rust/pull/116449#issuecomment-1807093362 is spurious or not.
@rustbot author

File: src/bootstrap/src/utils/helpers.rs (hunk 1 above)
Comment (Kobzol) [reply src/bootstrap/src/utils/helpers.rs:478]: Ok, renamed.

**Comment by Kobzol** [comment]
I'm also unsure about the CI failure, it happened repeatedly, but I can't reproduce it locally. Let's see what happens now.

**Comment by Kobzol** [comment]
There was a problem with empty `RUSTDOCFLAGS`. If I passed the env. var. empty, one rustdoc-gui test failed. So I added a check to skip setting the environment variable if it's empty.

@rustbot ready

**Comment by Kobzol** [comment]
A redundant use statement was in windows `cfg`-gated code.

@bors r=petrochenkov

**Pull request merged**